addons/website_slides/models/slide_channel.py

```diff
@@ -661,63 +789,118 @@ def action_redirect_to_members(self, completed=False):
+
+        If the logged in user has been invited, then we sudo the add_members method
+        since we consider he has been granted access to the course.
+
+        Returns True if added and enrolled successfully, False otherwise."""
+        invited_res = False
+        courses_invited = self.filtered(lambda course: course.is_member_invited)
+        if courses_invited:
+            invited_res = bool(courses_invited.sudo()._action_add_members(self.env.user.partner_id))
+
+        return invited_res or bool((self - courses_invited)._action_add_members(self.env.user.partner_id))
+
+    def _action_add_members(self, target_partners, member_status='joined'):
+        """ Adds the target_partners as attendees of the channel(s).
+            Partners are added as follows, depending on the value of member_status:
+            1) (Default) 'joined'. The partners will be added as enrolled attendees. This will make the content
+                (slides) of the channel available to that partner. This can also happen when an invited attendee
+                enrolls themself. The attendes are also subscribed to the chatter of the channel.
+            2) 'invited' : This is used when inviting partners. The partners are added as invited attendees
+                This will make the channel accessible but not the slides until they enroll themselves.
+
+        Returns the union of new partners and of invited partners having enrolled themselves."""
+        is_invite = member_status == 'invited'
+
+        # Channels on which the current user has the rights to write.
+        channels_to_join = self._filter_add_members(target_partners)
+
+        # Existing channel partners
+        existing_channel_partners = self.env['slide.channel.partner'].sudo().search([
+            ('channel_id', 'in', channels_to_join.ids),
+            ('partner_id', 'in', target_partners.ids)
+        ])
+        existing_channel_partners_map = defaultdict(list)
+        for channel_partner in existing_channel_partners:
+            existing_channel_partners_map[channel_partner.channel_id.id].append(channel_partner.partner_id.id)
+
+        # Invited partners confirming their invitation by enrolling to the course.
+        confirmed_invitation_partners_sudo = self.env['slide.channel.partner'].sudo()
+        to_create_partners_values = []
+
+        for channel in channels_to_join:
+            for partner in target_partners:
+                if partner.id not in existing_channel_partners_map[channel.id]:
+                    to_create_partners_values.append(dict(channel_id=channel.id, partner_id=partner.id, member_status=member_status))
+                elif not is_invite:
+                    existing_partner = existing_channel_partners.filtered(lambda cp: cp.partner_id.id == partner.id and cp.channel_id.id == channel.id)
+                    if existing_partner.member_status == 'invited':
+                        confirmed_invitation_partners_sudo |= existing_partner
```

Comment:
This was difficult to quite difficult to read and understand.
* The first thing that could help is using a defaultdict(set) instead of list, it will improve the various `in` checks.
* It you do a mapping of int => int[] or {...ints}, which is valid, you could then loop over ids like `for channel_id in channels_to_join.ids` etc
* But actually you need to handle more things from the records, so your map is not enough and you go back to the recordset `existing_channel_partners`.
* You have a nested loop for channel partners, and you skip some opportunities offered by the framework.

I'd like to suggest two main changes:
* Make a `slide.channel` => `slide.channel.partner`s mapping
* Use the ~~force~~ ORM and more recordsets instead of invidual records, dropping one of the loops

Some items changed names, all in all I think this it simpler:

```suggestion
        existing_channel_partners_map = defaultdict(lambda _:  self.env['slide.channel.partner'])
        for channel_partner in existing_channel_partners:
            existing_channel_partners_map[channel_partner.channel_id] |= channel_partner.partner_id

        channel_partners_joined_sudo = self.env["slide.channel.partner"].sudo()
        to_create_partners_values = []

        for channel_id, channel_partners in existing_channel_partners_map.items():
            if member_status != 'invited':
                channel_partners_joined_sudo |= channel_partners.filtered(
                    lambda cp: cp.member_status == 'invited')

            missing_channel_partners = target_partners.filtered(lambda p: p not in channel_partners.partner_id.ids)
            for partner_id in missing_channel_partners:
                to_create_partners_values.append(dict(channel_id=channel_id, partner_id=partner_id, member_status=member_status))

```

Reply:
Good point, a nice use of defaultdict that I did not think of. A few changes to your suggested change and it seems to work just fine, is clearer / shorter.

Main change is that I cannot use items() because it relies on existing_channel_partners: for instance if no partners exist, nothing will be created and the loop will not be entered. Instead I loop on every channel and access directly their attendees in the default dict, so that every one of them is dealt with.

Small conditions can optimize the code a little but I've kept it short and clean. Let me know if we should filter the target partners only if channel_partners is set, or if we prefer the code as is ;)